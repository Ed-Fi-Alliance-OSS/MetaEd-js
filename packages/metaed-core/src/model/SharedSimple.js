// @flow
import { newNamespaceInfo } from './NamespaceInfo';
import { ModelBase } from './ModelBase';
import type { SourceMap } from './SourceMap';

export class SharedSimpleSourceMap {
  type: ?SourceMap;
  documentation: ?SourceMap;
  metaEdName: ?SourceMap;
  metaEdId: ?SourceMap;
  namespaceInfo: ?SourceMap;
}

export class SharedSimple extends ModelBase {
  typeHumanizedName: string;
  sourceMap: SharedSimpleSourceMap;
}

export function defaultSharedSimple(): SharedSimple {
  return Object.assign(new SharedSimple(), {
    type: 'unknown',
    typeHumanizedName: 'unknown',
    documentation: '',
    metaEdName: '',
    metaEdId: '',
    namespaceInfo: newNamespaceInfo(),
    sourceMap: new SharedSimpleSourceMap(),
  });
}

export const NoSharedSimple: SharedSimple = Object.assign(defaultSharedSimple(), {
  metaEdName: 'NoSharedSimple',
});
