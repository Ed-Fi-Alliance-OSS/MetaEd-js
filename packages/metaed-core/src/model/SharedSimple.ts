// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import deepFreeze from 'deep-freeze';
import { NoNamespace } from './Namespace';
import { ModelBase, newModelBase, ModelBaseSourceMap, newModelBaseSourceMap } from './ModelBase';
import { SimpleProperty } from './property/SimpleProperty';

/**
 *
 */
export type SharedSimpleSourceMap = ModelBaseSourceMap;

/**
 *
 */
export function newSharedSimpleSourceMap(): SharedSimpleSourceMap {
  return newModelBaseSourceMap();
}

export interface SharedSimple extends ModelBase {
  typeHumanizedName: string;
  inReferences: SimpleProperty[];
  sourceMap: SharedSimpleSourceMap;
}

/**
 *
 */
export function newSharedSimple(): SharedSimple {
  return {
    ...newModelBase(),
    typeHumanizedName: 'unknown',
    namespace: NoNamespace,
    inReferences: [],
    sourceMap: newSharedSimpleSourceMap(),
  };
}

/**
 *
 */
export const NoSharedSimple: SharedSimple = deepFreeze({
  ...newSharedSimple(),
  metaEdName: 'NoSharedSimple',
});
