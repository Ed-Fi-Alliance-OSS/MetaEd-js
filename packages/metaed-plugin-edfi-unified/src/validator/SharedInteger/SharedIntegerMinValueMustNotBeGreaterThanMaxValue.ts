import { MetaEdEnvironment, ValidationFailure, SharedIntegerSourceMap, Namespace } from '@edfi/metaed-core';

// @ts-ignore
export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.namespace.forEach((namespace: Namespace) => {
    namespace.entity.sharedInteger.forEach((entity) => {
      if (
        entity.minValue &&
        entity.maxValue &&
        Number.parseInt(entity.minValue, 10) > Number.parseInt(entity.maxValue, 10)
      ) {
        failures.push({
          validatorName: 'SharedIntegerMinValueMustNotBeGreaterThanMaxValue',
          category: 'error',
          message: `${entity.typeHumanizedName} ${entity.metaEdName} has min value greater than max value.`,
          sourceMap: (entity.sourceMap as SharedIntegerSourceMap).minValue,
          fileMap: null,
        });
      }
    });
  });

  return failures;
}
