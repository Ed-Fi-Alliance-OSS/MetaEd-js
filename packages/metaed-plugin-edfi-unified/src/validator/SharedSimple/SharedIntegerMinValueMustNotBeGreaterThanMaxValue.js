// @flow
import type { MetaEdEnvironment, ValidationFailure, SharedIntegerSourceMap } from 'metaed-core';

// eslint-disable-next-line no-unused-vars
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.entity.sharedInteger.forEach(entity => {
    if (entity.minValue && entity.maxValue && Number.parseInt(entity.minValue, 10) > Number.parseInt(entity.maxValue, 10)) {
      failures.push({
        validatorName: 'SharedIntegerMinValueMustNotBeGreaterThanMaxValue',
        category: 'error',
        message: `${entity.typeHumanizedName} ${entity.metaEdName} has min value greater than max value.`,
        sourceMap: ((entity.sourceMap: any): SharedIntegerSourceMap).minValue,
        fileMap: null,
      });
    }
  });

  return failures;
}
