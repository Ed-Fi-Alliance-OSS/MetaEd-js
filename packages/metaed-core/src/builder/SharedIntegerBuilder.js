// @flow
import type { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { SharedSimpleBuilder } from './SharedSimpleBuilder';
import { newSharedInteger } from '../model/SharedInteger';
import type { SharedInteger, SharedIntegerSourceMap } from '../model/SharedInteger';
import { sourceMapFrom } from '../model/SourceMap';
import { isErrorText } from './BuilderUtility';
import { NoSharedSimple } from '../model/SharedSimple';

export class SharedIntegerBuilder extends SharedSimpleBuilder {
  enterSharedInteger(context: MetaEdGrammar.SharedIntegerContext) {
    this.enteringSharedSimple(newSharedInteger);
    if (this.currentSharedSimple !== NoSharedSimple) {
      Object.assign(((this.currentSharedSimple.sourceMap: any): SharedIntegerSourceMap), {
        type: sourceMapFrom(context),
        namespaceInfo: this.currentSharedSimple.namespaceInfo.sourceMap.type,
      });
    }
  }

  enterSharedShort(context: MetaEdGrammar.SharedShortContext) {
    this.enteringSharedSimple(newSharedInteger);
    if (this.currentSharedSimple !== NoSharedSimple) {
      ((this.currentSharedSimple: any): SharedInteger).isShort = true;
      Object.assign(((this.currentSharedSimple.sourceMap: any): SharedIntegerSourceMap), {
        type: sourceMapFrom(context),
        isShort: sourceMapFrom(context),
        namespaceInfo: this.currentSharedSimple.namespaceInfo.sourceMap.type,
      });
    }
  }

  // eslint-disable-next-line no-unused-vars
  exitSharedInteger(context: MetaEdGrammar.SharedIntegerContext) {
    this.exitingSharedSimple();
  }

  // eslint-disable-next-line no-unused-vars
  exitSharedShort(context: MetaEdGrammar.SharedShortContext) {
    this.exitingSharedSimple();
  }

  enterSharedIntegerName(context: MetaEdGrammar.SharedIntegerNameContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
    ((this.currentSharedSimple.sourceMap: any): SharedIntegerSourceMap).metaEdName = sourceMapFrom(context);
  }

  enterSharedShortName(context: MetaEdGrammar.SharedShortNameContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
    ((this.currentSharedSimple.sourceMap: any): SharedIntegerSourceMap).metaEdName = sourceMapFrom(context);
  }

  enterMinValue(context: MetaEdGrammar.MinValueContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;
    if (context.exception || context.signed_int() == null || context.signed_int().exception || isErrorText(context.signed_int().getText())) return;
    ((this.currentSharedSimple: any): SharedInteger).minValue = context.signed_int().getText();
    ((this.currentSharedSimple.sourceMap: any): SharedIntegerSourceMap).minValue = sourceMapFrom(context);
  }

  enterMaxValue(context: MetaEdGrammar.MaxValueContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;
    if (context.exception || context.signed_int() == null || context.signed_int().exception || isErrorText(context.signed_int().getText())) return;
    ((this.currentSharedSimple: any): SharedInteger).maxValue = context.signed_int().getText();
    ((this.currentSharedSimple.sourceMap: any): SharedIntegerSourceMap).maxValue = sourceMapFrom(context);
  }
}
