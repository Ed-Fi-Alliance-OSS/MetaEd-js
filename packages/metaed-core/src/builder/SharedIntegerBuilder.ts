// SPDX-License-Identifier: Apache-2.0
// Licensed to the Ed-Fi Alliance under one or more agreements.
// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
// See the LICENSE and NOTICES files in the project root for more information.

import { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { SharedSimpleBuilder } from './SharedSimpleBuilder';
import { newSharedInteger } from '../model/SharedInteger';
import { SharedInteger, SharedIntegerSourceMap } from '../model/SharedInteger';
import { sourceMapFrom } from '../model/SourceMap';
import { isErrorText } from './BuilderUtility';
import { NoSharedSimple } from '../model/SharedSimple';

/**
 * An ANTLR4 listener that creates SharedInteger entities.
 */
export class SharedIntegerBuilder extends SharedSimpleBuilder {
  enterSharedInteger(context: MetaEdGrammar.SharedIntegerContext) {
    this.enteringSharedSimple(newSharedInteger);
    if (this.currentSharedSimple !== NoSharedSimple) {
      Object.assign(this.currentSharedSimple.sourceMap as SharedIntegerSourceMap, {
        type: sourceMapFrom(context),
      });
    }
  }

  enterSharedShort(context: MetaEdGrammar.SharedShortContext) {
    this.enteringSharedSimple(newSharedInteger);
    if (this.currentSharedSimple !== NoSharedSimple) {
      (this.currentSharedSimple as SharedInteger).isShort = true;
      Object.assign(this.currentSharedSimple.sourceMap as SharedIntegerSourceMap, {
        type: sourceMapFrom(context),
        isShort: sourceMapFrom(context),
      });
    }
  }

  exitSharedInteger(_context: MetaEdGrammar.SharedIntegerContext) {
    this.exitingSharedSimple();
  }

  exitSharedShort(_context: MetaEdGrammar.SharedShortContext) {
    this.exitingSharedSimple();
  }

  enterSharedIntegerName(context: MetaEdGrammar.SharedIntegerNameContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
    (this.currentSharedSimple.sourceMap as SharedIntegerSourceMap).metaEdName = sourceMapFrom(context);
  }

  enterSharedShortName(context: MetaEdGrammar.SharedShortNameContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
    (this.currentSharedSimple.sourceMap as SharedIntegerSourceMap).metaEdName = sourceMapFrom(context);
  }

  enterMinValue(context: MetaEdGrammar.MinValueContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;

    if (context.exception) return;
    if (context.signed_int() == null && context.BIG() == null) return;
    if (context.signed_int() == null && (context.BIG().exception || isErrorText(context.BIG().getText()))) return;
    if (context.BIG() == null && (context.signed_int().exception || isErrorText(context.signed_int().getText()))) return;

    if (context.signed_int() != null) (this.currentSharedSimple as SharedInteger).minValue = context.signed_int().getText();
    if (context.BIG() != null) (this.currentSharedSimple as SharedInteger).hasBigHint = true;

    (this.currentSharedSimple.sourceMap as SharedIntegerSourceMap).minValue = sourceMapFrom(context);
  }

  enterMaxValue(context: MetaEdGrammar.MaxValueContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;

    if (context.exception) return;
    if (context.signed_int() == null && context.BIG() == null) return;
    if (context.signed_int() == null && (context.BIG().exception || isErrorText(context.BIG().getText()))) return;
    if (context.BIG() == null && (context.signed_int().exception || isErrorText(context.signed_int().getText()))) return;

    if (context.signed_int() != null) (this.currentSharedSimple as SharedInteger).maxValue = context.signed_int().getText();
    if (context.BIG() != null) (this.currentSharedSimple as SharedInteger).hasBigHint = true;

    (this.currentSharedSimple.sourceMap as SharedIntegerSourceMap).maxValue = sourceMapFrom(context);
  }
}
