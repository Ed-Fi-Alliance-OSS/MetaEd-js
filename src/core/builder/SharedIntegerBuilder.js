// @flow
import { MetaEdGrammar } from '../../grammar/gen/MetaEdGrammar';
import SharedSimpleBuilder from './SharedSimpleBuilder';
import { sharedIntegerFactory } from '../model/SharedInteger';
import type { SharedInteger } from '../model/SharedInteger';
import { isErrorText } from './BuilderUtility';
import { NoSharedSimple } from '../model/SharedSimple';

export default class SharedIntegerBuilder extends SharedSimpleBuilder {
  // eslint-disable-next-line no-unused-vars
  enterSharedInteger(context: MetaEdGrammar.SharedIntegerContext) {
    this.enteringSharedSimple(sharedIntegerFactory);
  }

  // eslint-disable-next-line no-unused-vars
  enterSharedShort(context: MetaEdGrammar.SharedShortContext) {
    this.enteringSharedSimple(sharedIntegerFactory);
    ((this.currentSharedSimple: any): SharedInteger).isShort = true;
  }

  // eslint-disable-next-line no-unused-vars
  exitSharedInteger(context: MetaEdGrammar.SharedIntegerContext) {
    this.exitingSharedSimple();
  }

  // eslint-disable-next-line no-unused-vars
  exitSharedShort(context: MetaEdGrammar.SharedShortContext) {
    this.exitingSharedSimple();
  }

  enterSharedIntegerName(context: MetaEdGrammar.SharedIntegerNameContext) {
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
  }

  enterSharedShortName(context: MetaEdGrammar.SharedShortNameContext) {
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
  }

  enterMinValue(context: MetaEdGrammar.MinValueContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;
    if (context.exception || context.signed_int() == null || context.signed_int().exception || isErrorText(context.signed_int().getText())) return;
    ((this.currentSharedSimple: any): SharedInteger).minValue = context.signed_int().getText();
  }

  enterMaxValue(context: MetaEdGrammar.MaxValueContext) {
    if (this.currentSharedSimple === NoSharedSimple) return;
    if (context.exception || context.signed_int() == null || context.signed_int().exception || isErrorText(context.signed_int().getText())) return;
    ((this.currentSharedSimple: any): SharedInteger).maxValue = context.signed_int().getText();
  }
}
