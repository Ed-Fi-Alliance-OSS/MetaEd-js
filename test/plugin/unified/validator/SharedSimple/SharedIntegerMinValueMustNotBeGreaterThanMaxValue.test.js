// @flow
import SharedIntegerBuilder from '../../../../../src/core/builder/SharedIntegerBuilder';
import MetaEdTextBuilder from '../../../../core/MetaEdTextBuilder';
import { repositoryFactory } from '../../../../../src/core/model/Repository';
import type { Repository } from '../../../../../src/core/model/Repository';
import { validate } from '../../../../../src/plugin/unified/validator/SharedSimple/SharedIntegerMinValueMustNotBeGreaterThanMaxValue';
import type { ValidationFailure } from '../../../../../src/core/validator/ValidationFailure';

describe('when validating shared integer with max value greater than min value', () => {
  const repository: Repository = repositoryFactory();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartSharedInteger('EntityName')
      .withDocumentation('PropertyDocumentation')
      .withMinValue('10')
      .withMaxValue('100')
      .withEndSharedInteger()
      .withEndNamespace()
      .sendToListener(new SharedIntegerBuilder(repository.entity, []));

    failures = validate(repository);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating shared integer with min value greater than max value', () => {
  const repository: Repository = repositoryFactory();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartSharedInteger('EntityName')
      .withDocumentation('PropertyDocumentation')
      .withMinValue('100')
      .withMaxValue('10')
      .withEndSharedInteger()
      .withEndNamespace()
      .sendToListener(new SharedIntegerBuilder(repository.entity, []));

    failures = validate(repository);
  });

  it('should build one shared integer', () => {
    expect(repository.entity.sharedInteger.size).toBe(1);
  });

  it('should have validation failures', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('SharedIntegerMinValueMustNotBeGreaterThanMaxValue');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when validating shared integer with min value greater than max value -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when validating shared integer with min value greater than max value -> sourceMap');
  });
});

describe('when validating shared short with max value greater than min value', () => {
  const repository: Repository = repositoryFactory();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartSharedShort('EntityName')
      .withDocumentation('PropertyDocumentation')
      .withMinValue('10')
      .withMaxValue('100')
      .withEndSharedShort()
      .withEndNamespace()
      .sendToListener(new SharedIntegerBuilder(repository.entity, []));

    failures = validate(repository);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating shared short with min value greater than max value', () => {
  const repository: Repository = repositoryFactory();
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartSharedShort('EntityName')
      .withDocumentation('PropertyDocumentation')
      .withMinValue('100')
      .withMaxValue('10')
      .withEndSharedShort()
      .withEndNamespace()
      .sendToListener(new SharedIntegerBuilder(repository.entity, []));

    failures = validate(repository);
  });

  it('should build one shared integer', () => {
    expect(repository.entity.sharedInteger.size).toBe(1);
  });

  it('should have validation failures', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('SharedIntegerMinValueMustNotBeGreaterThanMaxValue');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when validating shared short with min value greater than max value -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when validating shared short with min value greater than max value -> sourceMap');
  });
});
