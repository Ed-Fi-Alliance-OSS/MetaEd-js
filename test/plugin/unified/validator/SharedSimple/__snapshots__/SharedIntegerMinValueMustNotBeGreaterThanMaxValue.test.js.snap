// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating shared integer with min value greater than max value -> message 1`] = `"Shared Integer EntityName has min value greater than max value."`;

exports[`when validating shared integer with min value greater than max value -> sourceMap 1`] = `
Object {
  "column": 4,
  "line": 5,
  "tokenText": "min value",
}
`;

exports[`when validating shared short with min value greater than max value -> message 1`] = `"Shared Integer EntityName has min value greater than max value."`;

exports[`when validating shared short with min value greater than max value -> sourceMap 1`] = `
Object {
  "column": 4,
  "line": 5,
  "tokenText": "min value",
}
`;
